R/tar.R

```diff
@@ -0,0 +1,215 @@
+
+#' Create a tar background process
+#'
+#' Use an external tar program, if there is a working one, otherwise use
+#' the internal implementation.
+#'
+#' When using the internal implementation, we need to start another R
+#' process.
+#'
+#' @param tarfile Tar file.
+#' @param files Files or regular expressions to set what to extract. if
+#'   `NULL` then everything is extracted.
+#' @param exdir Where to extract the archive. It must exist.
+#' @param restore_times Whether to restore file modification times.
+#' @param post_process Function to call after the extraction.
+#' @return The [callr::process] object.
+#' @keywords internal
+
+make_untar_process <- function(tarfile, files = NULL, exdir = ".",
+                               restore_times = TRUE, post_process = NULL) {
+  internal <- need_internal_tar()
+  if (internal) {
+    r_untar_process$new(tarfile, files, exdir, restore_times,
+                        post_process = post_process)
+  } else {
+    external_untar_process$new(tarfile, files, exdir, restore_times,
+                               post_process = post_process)
+  }
+}
+
+#' Check if we need to use R's internal tar implementation
+#'
+#' This is slow, because we need to start an R child process, and the
+#' implementation is also very slow. So it is better to use an extranl tar
+#' program, if we can. We test this by trying to uncompress a .tar.gz
+#' archive using the external program. The name of the tar program is
+#' taken from the `TAR` environment variable, if this is unset then `tar`
+#' is used.
+#'
+#' @return Whether we need to use the internal tar implementation.
+#' @keywords internal
+
+need_internal_tar <- (function() {
```

Comment:
I think it is more personally more clear (at least to me) to use a `local()` block then a closure you immediately call.

Reply:
ok